modules/vebal/root-gauges.onchain.ts

```diff
@@ -0,0 +1,221 @@
+
+import { mainnetNetworkConfig } from '../network/mainnet';
+import { gaugeControllerAbi } from './abi/gaugeController.abi';
+import { rootGaugeAbi } from './abi/rootGauge.abi';
+
+const gaugeControllerContract = {
+    address: mainnetNetworkConfig.data.gaugeControllerAddress!,
+    abi: gaugeControllerAbi,
+} as const;
+
+export type RootGauge = {
+    gaugeAddress: Address;
+    network: Chain;
+    isKilled: boolean;
+    relativeWeight: number;
+    relativeWeightCap?: string;
+    recipient?: string;
+    stakingId?: Address;
+    isInSubgraph: boolean;
+};
+
+/**
+ *
+ * This is the first proof of concept using viem's multicall
+ * It contains ad-hoc helpers to make the code easier in this concrete scenario but in the future we will try to create a more generic
+ * multicaller implementation to generalize any multicall flow
+ *
+ */
+export class OnChainRootGauges {
+    constructor(
+        private publicClient: PublicClient = mainnetNetworkConfig.publicClient!,
+        private readContract = publicClient.readContract,
+    ) {}
+
+    async getRootGaugeAddresses(): Promise<string[]> {
+        const totalGauges = Number(
+            await this.readContract({
+                ...gaugeControllerContract,
+                functionName: 'n_gauges',
+            }),
+        );
+        const addresses = await this.publicClient.multicall({
+            allowFailure: false,
+            contracts: this.gaugeControllerCallsByIndex(totalGauges, 'gauges'),
+        });
+
+        return addresses.map((address) => address.toLowerCase());
+    }
+
+    async fetchOnchainRootGauges(gaugeAddresses: Address[]): Promise<RootGauge[]> {
+        const totalGaugesTypes = Number(
+            await this.readContract({
+                ...gaugeControllerContract,
+                functionName: 'n_gauge_types',
+            }),
+        );
+
+        const typeNames = await this.publicClient.multicall({
+            allowFailure: false,
+            contracts: this.gaugeControllerCallsByIndex(totalGaugesTypes, 'gauge_type_names'),
+        });
+
+        const relativeWeights = await this.multicallGaugeController(gaugeAddresses, 'gauge_relative_weight');
+
+        /*
+            gauge_types are not reliable because they are manually input by Maxis
+            We will use subgraph chain field instead
+            However, we keep pulling this gauge_types cause they can be useful for debugging (when a root gauge is not found in the subgraph)
+        */
+        const gaugeTypeIndexes = await this.multicallGaugeController(gaugeAddresses, 'gauge_types');
+        const gaugeTypes = mapValues(gaugeTypeIndexes, (type) => typeNames[Number(type)]);
+
+        const isKilled = await this.multicallRootGauges(gaugeAddresses, 'is_killed');
+
+        const relativeWeightCapsWithFailures = await this.multicallRootGaugesAllowingFailures(
+            gaugeAddresses,
+            'getRelativeWeightCap',
+        );
```

Comment:
are there rootGauges that don't have this field? Or why do we need to allow failures?

Reply:
Currently 126 out of 333 root gauge deployed contracts do not have this field. Examples: 

```
  '0xf0ea3559cf098455921d74173da83ff2f6979495',
  '0x3fdb6fb126521a28f06893f9629da12f7b7266eb',
  '0xe0b50b0635b90f7021d2618f76ab9a31b92d0094',
  '0x785f08fb77ec934c01736e30546f87b4daccbe50',
  '0x981fb05b738e981ac532a99e77170ecb4bc27aef',
  '0xb0de49429fbb80c635432bbad0b3965b28560177',
  '0x359ea8618c405023fc4b98dab1b01f373792a126',
  '0x435272180a4125f3b47c92826f482fc6cc165958',
  '0x05e7732bf9ae5592e6aa05afe8cd80f7ab0a7bea',
  '0xc77e5645dbe48d54afc06655e39d3fe17eb76c1c',
  '0xf30db0ca4605e5115df91b56bd299564dca02666',
  '0x899f737750db562b88c1e412ee1902980d3a4844',
  '0x6cb1a77ab2e54d4560fda893e9c738ad770da0b0',
  '0x6823dca6d70061f2ae2aaa21661795a2294812bf',
  '0xa5a0b6598b90d214eaf4d7a6b72d5a89c3b9a72c',
  '0x88d07558470484c03d3bb44c3ecc36cafcf43253',
  '0xfbf87d2c22d1d298298ab5b0ec957583a2731d15',
```